### PR TITLE
[Fix](mluOpExecFFT): fix core dump, scale factor and one point compute error

### PR DESCRIPTION
**Pull request by aokbok**
Thanks for your contribution and we appreciate it a lot. :rocket::rocket:

## 1. Motivation

Please describe your motivation and the goal you want to achieve through this pull request.

## 2. Modification

Please briefly describe what modification is made in this pull request, and indicate where to make the modification.

Are new test cases added? If so, please post the corresponding generator-PR link here.

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS™ Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- static threshold
  - diff1
    - [ ] float32 mlu diff1 <= 1e-5
    - [ ] float32 mlu diff1 <= 3e-3
    - [ ] float16 mlu diff1 <= 3e-3
  - diff2
    - [ ] float32 mlu diff2 <= 1e-5
    - [ ] float32 mlu diff2 <= 3e-3
    - [ ] float16 mlu diff2 <= 3e-3
  - diff3
    - [ ] mlu diff3 == 0
    - [ ] mlu diff3_1 == 0
    - [ ] mlu diff3_2 == 0
- dynamic threshold
  - [ ] diff1: mlu diff1 <= max(baseline diff1 * 10, static threshold)
  - [ ] diff2: mlu diff2 <= max(baseline diff2 * 10, static threshold)
  - [ ] diff3: mlu diff3 <= max(baseline diff3 * 10, static threshold)
    - float32, threshold = 1e-5
    - float16, threshold = 1e-3

#### 3.1.2 Operator Scheme checklist

- Supported hardware
  - [ ] MLU370
  - [ ] MLU590
- Job types
  - [ ] BLOCK
  - [ ] UNION1
  - [ ] UNION2
  - [ ] UNION4
  - [ ] The operator will dynamically select the most suitable task type, for example, UNION8

#### 3.2.1 Accuracy Test

If you have checked the following items, please tick the relevant box.

- [ ] Data type test (e.g. float32/int8)
- [ ] Multi-dimensional tensor test
- [ ] Layout test
- [ ] Different size/integer remainder end segment/alignment misalignment test
- [ ] Zero dimensional tensor test/zero element test
- [ ] stability test
- [ ] Multiple platform test
- [ ] Gen_case module test, see: [Gencase-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/Gencase-User-Guide-zh.md)
- [ ] Nan/INF tests 
- [ ] Bug fix tests
- [ ] For memory leak check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [ ] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [ ] For I/O calculation efficiency check details, see: [MLU-OPS™-Performance-Acceptance-Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md)

#### 3.2.2 Parameter Check

Test Point-1: `When a new operator is submitted, the test points are given and the test results are stated`. Acceptance Standard: `Normal error`.
```bash
Please fill your test results(Error Message) in here, ...
```

Test Point-2: `Whether illegal parameters are passed`. Acceptance Standard: `Normal error`.
```bash
Test results...
```

### 3.3 Performance Test

See [MLU-OPS™ Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU370

```bash
# The test results should contain Op name, Shape, Data type,  
#   MLU Hardware Time(us), MLU Interface Time(us), MLU IO Efficiency, 
#   MLU Compute Efficiency, and Mlu Workspace Size(Bytes)
# 
# for example:
#
# ----------- case0 -----------
# case0
# [Op name                ]: abs
# [Shape                  ]: input.shape=[1024,1024,3,4], output.shape=[1024,1024,3,4]
# [Data type]             ]: float32
# [MLU Hardware Time      ]: 15728 (us)
# [MLU Interface Time     ]: 369.008 (us)
# [MLU IO Efficiency      ]: 0.23275
# [MLU Compute Efficiency ]: 0.5
# [Mlu Workspace Size     ]: -1 (Bytes)
# 
# ----------- case1 -----------
# ...
```

Platform：MLU590
```bash
# ----------- case0 -----------
# ----------- case1 -----------
# ...
```

### 3.4 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

### CODE REVIEW

**Comment by aokbok** [comment]
测试报告：
http://jenkins.svc.cambricon.com/dist/job/MLUOPS/job/MLU_ALL_PLATFORMS/job/mluops_build_test_adaptable_mlu/693/

**Comment by aokbok** [comment]
测试报告：
image

File: kernels/fft/c2c_fft/c2c_fft_host.cpp [hunk 1]
```diff
@@ -2053,7 +2058,40 @@ mluOpStatus_t execFFT2d(mluOpHandle_t handle, const mluOpFFTPlan_t fft_plan,
...
 
-  status = execFFTc2c2d(handle, fft_plan, scale_factor, direction);
+  if (fft_plan->n[0] == 1 && fft_plan->n[1] == 1) {
+    mluOpTensorDescriptor_t c_desc = nullptr;
+    status = mluOpCreateTensorDescriptor(&c_desc);
+    const int out_dim_num = 3;
+    int64_t dims[out_dim_num] = {fft_plan->batch, fft_plan->n[0],
+                                 fft_plan->n[1]};
+    status = mluOpSetTensorDescriptor_v2(c_desc, MLUOP_LAYOUT_ARRAY,
+                                         fft_plan->output_dtype, 2, dims);
```
Comment (zhengleiZL) [inline kernels/fft/c2c_fft/c2c_fft_host.cpp:2068]: dims是3维，但是创建的Tensor设置的是2

File: kernels/fft/c2c_fft/c2c_fft_host.cpp [hunk 2]
```diff
@@ -2053,7 +2058,40 @@ mluOpStatus_t execFFT2d(mluOpHandle_t handle, const mluOpFFTPlan_t fft_plan,
     fft_plan->mlu_addrs.input = fft_plan->mlu_addrs.input_pad_addr;
   }
 
-  status = execFFTc2c2d(handle, fft_plan, scale_factor, direction);
+  if (fft_plan->n[0] == 1 && fft_plan->n[1] == 1) {
+    mluOpTensorDescriptor_t c_desc = nullptr;
+    status = mluOpCreateTensorDescriptor(&c_desc);
```
Comment (Unireverse) [inline kernels/fft/c2c_fft/c2c_fft_host.cpp:2063]: c_desc好像没有destroy

File: kernels/fft/c2c_fft/c2c_fft_host.cpp [hunk 3]
```diff
@@ -2053,7 +2058,40 @@ mluOpStatus_t execFFT2d(mluOpHandle_t handle, const mluOpFFTPlan_t fft_plan,
...
+
+    DEFINE_CREATE_AND_SET_CNNL_TENSOR_DESCRIPTOR(c_desc, cnnl_output_desc);
+
+    size_t workspace_size = 0;
+    CALL_CNNL(cnnlGetCopyWorkspaceSize(cnnl_handle, cnnl_output_desc,
+                                       cnnl_output_desc, &workspace_size));
+    void *workspace = nullptr;
+    if (workspace_size > 0) {
+      // TODO(niyuming): ict code legacy issues, to be resolved in the future.
+      CNRT_CHECK(cnrtMalloc((void **)&workspace, workspace_size));
```
Comment (zhengleiZL) [inline kernels/fft/c2c_fft/c2c_fft_host.cpp:2083]: 要在算子里Malloc吗？不应该是调用算子之前就把workspace空间准备好吗？

也就是说在调用c2c_fft之前，先调用getC2CFFTWorkspace()函数，提前算出这块workspace。应该在etC2CFFTWorkspace()接口中增加这块workspace把

File: kernels/fft/irfft/irfft_host.cpp [hunk 4]
```diff
@@ -1987,63 +1991,87 @@ mluOpStatus_t execIRFFT2d(mluOpHandle_t handle, const mluOpFFTPlan_t fft_plan,
...
+    INTERNAL_CHECK(api, status == MLUOP_STATUS_SUCCESS);
+  }
 
-      DEFINE_CREATE_AND_SET_CNNL_HANDLE(handle,
-                                        cnnl_handle);  // convert to cnnl_handle
+  if (scale_factor != 1.0) {
+    const float alpha[2] = {scale_factor, 0.0};
+    const float beta[2] = {0.0, 0.0};
+    mluOpTensorDescriptor_t c_desc = nullptr;
+    status = mluOpCreateTensorDescriptor(&c_desc);
```
Comment (Unireverse) [inline kernels/fft/irfft/irfft_host.cpp:2048]: 同一个问题

File: kernels/fft/rfft/rfft_host.cpp [hunk 5]
```diff
@@ -1561,70 +1564,104 @@ mluOpStatus_t execRFFT2d(mluOpHandle_t handle, const mluOpFFTPlan_t fft_plan,
...
     INTERNAL_CHECK(api, status == MLUOP_STATUS_SUCCESS);
     status = computeFFT2dMatMulColumnR2C(handle, fft_plan, scale_factor);
     INTERNAL_CHECK(api, status == MLUOP_STATUS_SUCCESS);
   }
+
+  if (scale_factor != 1.0) {
+    const float alpha[2] = {scale_factor, 0.0};
+    const float beta[2] = {0.0, 0.0};
+    mluOpTensorDescriptor_t c_desc = nullptr;
+    status = mluOpCreateTensorDescriptor(&c_desc);
```
Comment (Unireverse) [inline kernels/fft/rfft/rfft_host.cpp:1639]: 这个也是

File: kernels/fft/fft_optm_device/fft_stockham_u1_device.mlu [hunk 6]
```diff
@@ -481,9 +481,9 @@ __mlu_func__ void store(DT* output, DT* y_in_r, DT* x_out1_r,
     if (part == 0) {
       int dst_one_point_offset = b * out_n * 2 + n;
       int src_one_point_offset = pow_2_m * L_sub;
-      __memcpy_async((DT *)output + dst_one_point_offset,
-                     (DT *)out_nram + src_one_point_offset, sizeof(DT) * 2,
-                     NRAM2GDRAM);
+      *(output + dst_one_point_offset) = *(out_nram + src_one_point_offset);
```
Comment (zhengleiZL) [inline kernels/fft/fft_optm_device/fft_stockham_u1_device.mlu:484]: 改成memcpy形式

File: kernels/fft/rfft/rfft_host.cpp [hunk 7]
```diff
@@ -1191,9 +1193,10 @@ static mluOpStatus_t makeRFFT2dContiguousOutput(mluOpHandle_t handle,
     const int out_dim_num = 3;
     int64_t dims[out_dim_num] = {fft_plan->batch, fft_plan->n[0],
                                  fft_plan->n[1] / 2 + 1};
-    int64_t strides[out_dim_num] = {fft_plan->odist,
-                                    fft_plan->ostride * fft_plan->onembed[1],
-                                    fft_plan->ostride};
+    int64_t strides[3];  // out_dim_num
```
Comment (zhengleiZL) [inline kernels/fft/rfft/rfft_host.cpp:1196]: 改成out_dim_num吧

File: kernels/fft/irfft/irfft_host.cpp [hunk 8]
```diff
@@ -1987,63 +1991,87 @@ mluOpStatus_t execIRFFT2d(mluOpHandle_t handle, const mluOpFFTPlan_t fft_plan,
...
+    const int out_dim_num = 3;
+    int64_t dims[out_dim_num] = {fft_plan->batch, fft_plan->n[0],
+                                 fft_plan->n[1]};
+    status = mluOpSetTensorDescriptor_v2(c_desc, MLUOP_LAYOUT_ARRAY,
+                                         fft_plan->output_dtype, 3, dims);
```
Comment (zhengleiZL) [inline kernels/fft/irfft/irfft_host.cpp:2053]: 在setTensor时候，3统一用out_dim_num这类变量吧，防止出错

发现有的时候用的是变量名，有的时候就直接用立即数

File: kernels/fft/fft.cpp [hunk 9]
```diff
@@ -1783,15 +1787,61 @@ mluOpAllocateC2R1D(mluOpHandle_t handle, mluOpFFTPlan_t fft_plan,
...
   }
   fft_plan->workspace_size = workspace_size;
   fft_plan->reservespace_size = reservespace_size;
 
   return MLUOP_STATUS_SUCCESS;
 }
+mluOpStatus_t MLUOP_WIN_API mluOpAllocateIRFFT2D(
+    mluOpHandle_t handle, mluOpFFTPlan_t fft_plan,
+    mluOpTensorDescriptor_t input_desc, mluOpTensorDescriptor_t output_desc,
+    const int _n0, const int _n1) {
```
Comment (PetrelYy) [inline kernels/fft/fft.cpp:1801]: _n0 为啥是这样的命名方式

**Review by PetrelYy — APPROVED** [review]

**Review by zhengleiZL — APPROVED** [review]

**Review by DanieeelLiu — APPROVED** [review]
LGTM

**Review by Unireverse — APPROVED** [review]
LGTM

**Pull request merged**